### PR TITLE
[WIP] Correct all the algos and benchmarking

### PR DESCRIPTION
**Pull request by Het-Shah**
I am going to add benchmarks for all the algorithms and will correct those that are not working.
- [x] VanillaKD
- [x] TAKD
- [ ] RKD
- [ ] noisy
- [ ] attention
- [ ] teacher free 
- [ ] mean teacher 
- [ ] RCO 
- [x] BANN 
- [ ] KA
- [ ] Bert2LSTM
- [x] DML

Feel free to take up any of the algorithms. Just comment on this pr which one are you taking.

### CODE REVIEW

**Comment by Het-Shah** [comment]
Did you guys not run the code to check if it is working or not? I am currently debugging every algo as i go about.

cc: @khizirsiddiqui

**Comment by Het-Shah** [comment]
Also, why are there two noisy in the README file? Please change that too, whoever has implemented them. It is confusing to use as of now.

**Comment by NeelayS** [comment]
> Did you guys not run the code to check if it is working or not? I am currently debugging every algo as i go about.
> 
> cc: @khizirsiddiqui

I'm pretty sure I did. Regardless, I'll take take up the ones I have implemented for benchmarking so that it'll be easy for me to debug if required - 

- noisy
- teacher_free
- RCO
- KA

**Comment by NeelayS** [comment]
> Also, why are there two noisy in the README file? Please change that too, whoever has implemented them. It is confusing to use as of now.
> 
> cc: @khizirsiddiqui @NeelayS

`noisy` currently has 2 paper implementations

**Comment by Het-Shah** [comment]
> noisy currently has 2 paper implementations

So both of them are in one folder itself?

**Comment by NeelayS** [comment]
Yup. Thought it would make sense

**Comment by NeelayS** [comment]
Should we benchmark on CIFAR-100 ? 

- Differences in performance will be visible better
- ResNet students will do well on MNIST without KD as well

**Comment by Het-Shah** [comment]
We can do both, I started with MNIST because it is faster to train.

**Comment by Het-Shah** [comment]
@khizirsiddiqui the implementations of attention and mean teacher need to be redone they are not just what you have implemented imo. Have a look at the implementations of these, they are quite different.

**Comment by Het-Shah** [comment]
RKD doesn't train too. There are errors in calculation of loss.

**Comment by NeelayS** [comment]
> > Did you guys not run the code to check if it is working or not? I am currently debugging every algo as i go about.
> > cc: @khizirsiddiqui
> 
> I'm pretty sure I did. Regardless, I'll take take up the ones I have implemented for benchmarking so that it'll be easy for me to debug if required -
> 
> * noisy
> * teacher_free
> * RCO
> * KA

I've tested all these algos here - https://colab.research.google.com/drive/1S3oluy7gA4LNwUH-tXWfsr-3vXS4sF9n?usp=sharing
I've also listed some issues and observations. 
@Het-Shah Can you go through this once

**Comment by Het-Shah** [comment]
There seems to be a lot of issues with the algos :/. Add the changes to this pr itself. Let's get all the current algos working before adding any new ones.

**Comment by Het-Shah** [comment]
Also, try using resnet models as well.

**Comment by NeelayS** [comment]
Except KA, all are training well but I can't figure out why they are performing worse than a model trained from scratch.

**Comment by Het-Shah** [comment]
We need to take care of imports properly as well. I think we should add relative imports everywhere to reduce the errors in imports at least. Also, we need to restructure the library itself into 3 major parts - Pruning, Quantization and KD.

**Comment by Het-Shah** [comment]
> Except KA, all are training well but I can't figure out why they are performing worse than a model trained from scratch.

This is a problem then we need to find the root cause of this maybe, otherwise KD doesn't make sense at all :/

**Comment by NeelayS** [comment]
Yeah exactly

**Comment by Het-Shah** [comment]
Maybe try for another dataset and see if it has the same pattern.

**Comment by NeelayS** [comment]
Should I try training for more epochs on this first ?

**Comment by Het-Shah** [comment]
Yea try for more epochs once, but MNIST will don't have a significant difference imo. Also, make sure what hyperparameters you use i.e. the temperature, distil_weight (if distil_weight is high then it is similar to training independently try keeping this to maybe 0.1/0.2) and other parameters as well depending on the algo.

**Comment by NeelayS** [comment]
**Issues -** 

- [ ] Knowledge Adjustment(KA) not training
- [ ] Issues in using CUDA with KA
- [ ] Import errors for RCO
- [ ] Poor performance for RCO(it's training though)
- [ ] Some algos work with nn.MSELoss() nut give extremely bad performance with nn.KLDivLoss() ad vice-versa
- [ ] None of the distilled models beat a Shallow_400 model(baseline) trained from scratch without using KD_Lib(trained at the bottom of the notebook)
- [ ] Very surprisingly, even the Shallow_1200 teachers seem to be performing worse than the Shallow_400 trained from scratch in some cases
- [ ] Inconsistent imports(compare DML vs others)
- [ ] Not really an issue, but the train accuracies for all the algos are lower than the test accuracies

**Comment by NeelayS** [comment]
> Yea try for more epochs once, but MNIST will don't have a significant difference imo. Also, make sure what hyperparameters you use i.e. the temperature, distil_weight (if distil_weight is high then it is similar to training independently try keeping this to maybe 0.1/0.2) and other parameters as well depending on the algo.

Trained on CIFAR10 for 20 epochs(in the same notebook). Still the same problem of performing worse than model without KD.
Will tune hyperparameters and try to check what's going wrong

**Comment by NeelayS** [comment]
> **Issues to solve -**
> 
> * [ ]  Knowledge Adjustment(KA) not training
> * [ ]  Issues in using CUDA with KA
> * [ ]  Import errors for RCO
> * [ ]  Poor performance for RCO(it's training though)
> * [ ]  Some algos work with nn.MSELoss() nut give extremely bad performance with nn.KLDivLoss() ad vice-versa
> * [x]  None of the distilled models beat a Shallow_400 model(baseline) trained from scratch without using KD_Lib(trained at the bottom of the notebook)
> * [x]  Very surprisingly, even the Shallow_1200 teachers seem to be performing worse than the Shallow_400 trained from scratch in some cases
> * [ ]  Inconsistent imports(compare DML vs others)
> * [ ]  Not really an issue, but the train accuracies for all the algos are lower than the test accuracies

All algos are beating the baseline. What happened is I unknowingly used plain SGD for the distillation process and SGD with momentum for training the baseline model. That's why the baseline model was performing better.

**Comment by NeelayS** [comment]
> There seems to be a lot of issues with the algos :/. Add the changes to this pr itself. Let's get all the current algos working before adding any new ones.

I don't have permission to push to this branch, do I ?

**Comment by Het-Shah** [comment]
You should have, wait let me check once though.

**Comment by Het-Shah** [comment]
Check once, I sent the invites again.

**Comment by Het-Shah** [comment]
Umm... you didn't pull before pushing? All my changes are gone :/

**Comment by Het-Shah** [comment]
Rebase to previous commit, pull then make changes and push.

**Comment by NeelayS** [comment]
> Rebase to previous commit, pull then make changes and push.

Yeah, sorry. I'll do that.

File: KD_Lib/noisy/soft_random.py [hunk 1]
```diff
@@ -154,8 +154,8 @@ def calculate_kd_loss(self, y_pred_student, y_pred_teacher, y_true):
         """
         loss = (1.0 - self.distil_weight) * F.cross_entropy(y_pred_student, y_true)
         loss += self.distil_weight * self.loss_fn(
-            F.log_softmax(y_pred_student / self.temp),
-            F.softmax(y_pred_teacher / self.temp),
+            F.log_softmax(y_pred_student / self.temp, dim=1),
+            F.softmax(y_pred_teacher / self.temp, dim=1),
```
Comment (Het-Shah) [inline KD_Lib/noisy/soft_random.py:158]: make  this `(self.distil_weight * self.temp * self.temp) * self.loss_fn(...)`. This generally gives faster convergence.

Comment (NeelayS) [reply KD_Lib/noisy/soft_random.py:158]: Okie

**Comment by NeelayS** [comment]
The teacher free methods do much better without the temp multiplier in the loss. I've removed it.

File: KD_Lib/KA/PS.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 from copy import deepcopy
 
 
-class KnowledgeAdjustment(BaseClass):
+class PS(BaseClass):
```
Comment (Het-Shah) [inline KD_Lib/KA/PS.py:12]: Can the class name be a bit more meaningful? I mean i won't know which implementation is this classname for?

File: logs.rst [hunk 3]
```diff
@@ -1,16 +1,58 @@
-Results while adding implementations -
-===========================
+`Colab Notebook <https://colab.research.google.com/drive/1brU6xL8nqlw5F6op0Edys-Av7SlTrr6t?usp=sharing>`_
```
Comment (Het-Shah) [inline logs.rst:1]: I guess we can make one example notebook later no need to have this in the `logs.rst` file as of now.

Comment (NeelayS) [reply logs.rst]: Removed

File: KD_Lib/KA/PS.py (hunk 2 above)
Comment (NeelayS) [reply KD_Lib/KA/PS.py]: Done

**Comment by NeelayS** [comment]
> > **Issues to solve -**
> > 
> > * [ ]   Knowledge Adjustment(KA) not training
> > * [ ]   Issues in using CUDA with KA

The issues with KA are still there, I'm not really sure what's going wrong there.
I was thinking this PR could be merged and we'll track KA in another issue ?

**Comment by Het-Shah** [comment]
> The issues with KA are still there, I'm not really sure what's going wrong there.
I was thinking this PR could be merged and we'll track KA in another issue ?

Yea sure I think we have done a lot of work in this PR anyway. I will merge it once the build completes.

**Comment by Het-Shah** [comment]
> I am going to add benchmarks for all the algorithms and will correct those that are not working.
> 
> * [x]  VanillaKD
> * [x]  TAKD
> * [ ]  RKD
> * [ ]  noisy
> * [ ]  attention
> * [ ]  teacher free
> * [ ]  mean teacher
> * [ ]  RCO
> * [x]  BANN
> * [ ]  KA
> * [ ]  Bert2LSTM
> * [x]  DML
> 
> Feel free to take up any of the algorithms. Just comment on this pr which one are you taking.

So how many are we done with?

**Comment by NeelayS** [comment]
- [ ] Attention
- [x] BANN
- [ ] BERT2LSTM
- [x] DML
- [ ] KA - LSR
- [ ] KA - PS
- [ ] Mean Teacher
- [x] Noisy - Noisy Teacher
- [x] Noisy - Soft Random
- [x] Noisy - Messy Collab
- [x] Pruning
- [x] RCO
- [ ] RKD
- [x] TAKD
- [x] Teacher Free - Self Training
- [x] Teacher Free - Virtual Teacher
- [x] Vanilla KD

**Comment by Het-Shah** [comment]
Great so i guess we will open up an issue with the remaining ones. So I am merging this pr as of now and opening an issue for the remaining ones.
